[experiment] Move old assumptions from Basic to Expr

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

This is still an experiment at this stage. The old assumptions have no place being on Basic rather than Expr and this PR is to see what happens if they are moved to Expr.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

asmeurer commented on sympy/polys/polytools.py:
```diff
@@ -5279,10 +5278,11 @@ def gcd(f, g=None, *gens, **args):
-            frc = (a/b).ratsimp()
-            if frc.is_rational:
-                return a/frc.as_numer_denom()[0]
+        if isinstance(a, Expr) and isinstance(b, Expr):
```
Is the sympify() line above supposed to ensure this? Otherwise if this code used to work on some polys objects we should add is_algebraic and is_irrational to them.

oscarbenjamin replied:
`Poly.is_algebraic/irrational` are always None on master.
```julia
In [1]: print(Poly(1, x).is_irrational)                                                                                                                       
None
```
In this PR `Poly` does not have those attributes.

The use of `as_expr` here ensures that we get an `Expr` whether the input was `Expr` or `Poly`. It would be a useful bugfix regardless of the other changes here.

Note that I didn't write this expecting to merge this PR (yet). I don't think that the old assumptions should be on Basic rather than Expr but this is more of an experiment. However now that the tests have passed I can say that every example where code needed to be changed here shows a bug in sympy which leads to two conclusions:

1. Something similar to each of the fixes here should be applied to fix bugs on master.

2. Having the old assumptions on Basic hides bugs. They should be moved to Expr only.

Given those maybe I should flesh this PR out to something mergeable but I haven't discussed with anyone the idea of moving the old assumptions yet.

asmeurer replied:
Is the None thing also true for the other classes? It looks like adding `isinstance(expr, Expr)` to all the conditions would change the logic, but if they assumptions would be None otherwise maybe the logic hasn't changed after all.

oscarbenjamin replied:
It doesn't change the logic because these properties are not meaningful for most non-Expr classes and so are already None in almost all cases.

Poly now subclasses Basic and never did have any `_eval_is` methods that would be needed to make these non-None:
```console
$ git grep '_eval_is' sympy/polys/
sympy/polys/rootoftools.py:    def _eval_is_real(self):
sympy/polys/rootoftools.py:    def _eval_is_imaginary(self):
```
Those are methods of `CRootOf` (which is `Expr`). There are none of these on `Poly` or any of the other classes in `polys`.

I made some changes to `ManagedProperties` to print out which non-`Expr` classes defined any of the assumptions related properties as static attributes and the only ones were cases of `is_commutative = True` (which are removed in this PR). I haven't systematically checked for `_eval_is` methods in non-`Expr` `Basic` subclasses: there aren't any that are tested as this PR shows.

eric-wieser commented on sympy/core/assumptions.py:
```diff
@@ -543,7 +568,14 @@ def __init__(cls, *args, **kws):
 
         cls._explicit_class_assumptions = defs
-        cls.default_assumptions = StdFactKB(defs)
+        if 'Expr' not in [c.__name__ for c in cls.mro()]:
```
```suggestion
        from sympy.core.expr import Expr
        if not isinstance(c, Expr):
```

oscarbenjamin replied:
There should be a better mechanism for this rather than special casing Expr here.

oscarbenjamin replied:
I think it would be better to use a separate metaclass for Expr e.g.:
```python
class BasicMeta(type):
    """Root metaclass for all Basic"""
    pass

class FactKB:
    """Class that handles logic inference for assumption systems"""
    pass

class EmptyKB(FactKB):
    """Logic inference system with no assumptions predicates"""
    pass

class ExprFactKB(FactKB):
    """Assumptions system for Expr with standard predicates"""
    pass

class BasicManagedProperties(BasicMeta):
    """Actual metaclass for Basic"""
    _FACTKB = EmptyKB

# The metaclass of a subclass must be a subclass of the metaclass of the
# superclass. This means that if Expr is a subclass of Basic then the metaclass
# for Expr must be a subclass of the metaclass for Basic.
class ExprManagedProperties(BasicManagedProperties):
    """Actual metaclass for Expr"""
    _FACTKB = ExprFactKB

class Basic(metaclass=BasicManagedProperties):
    """Root superclass for all symbolic things"""
    pass

class Expr(Basic, metaclass=ExprManagedProperties):
    """Root superclass for all expression-like things"""
    pass

```

eric-wieser commented on sympy/core/assumptions.py:
```diff
@@ -569,7 +601,8 @@ def __init__(cls, *args, **kws):
                 setattr(cls, pname, make_property(fact))
 
         # Finally, add any missing automagic property (e.g. for Basic)
```
```suggestion
        # Finally, add any missing automagic property (e.g. for Expr)
```

oscarbenjamin committed:
core: move old assumptions from Basic to Expr

The old assumptions were added to all Basic instances through the
ManagedProperties metaclass. This commit modifies the metaclass so that
it only adds the old assumptions properties to subclasses of Expr rather
than all of Basic.

czgdp1807 commented:
@oscarbenjamin Is the purpose of this PR fulfilled?

oscarbenjamin commented:
This should still be completed. What is missing is a deprecated compatibility assumptions system for non-Expr Basic subclasses.

asmeurer commented:
Shouldn't the core change here be moving the `ManagedProperties` metaclass from `Basic` to `Expr`? What's the function of the `EmptyKB`?

I'm also curious how this interacts with the new assumptions. The new assumptions are allowed on any kind of class (e.g., there are some matrix assumptions). Does the code in the new assumptions that calls the old assumptions need to be updated?

oscarbenjamin commented:
> What's the function of the `EmptyKB`?

The `EmptyKB` here doesn't really do anything. In principle though we could make different FactKBs for different types of object e.g. sets, matrices etc. Then the EmptyKB would just be the fallback for types that don't have any associated predicates.

Alternatively the `EmptyKB` here could be made into something that emits deprecation warnings any time someone tries to use the old assumptions with a non-Expr class.

> Does the code in the new assumptions that calls the old assumptions need to be updated?

I don't think so:
```python
In [6]: s = FiniteSet(1)

In [7]: ask(Q.zero(x))

In [8]: print(ask(Q.zero(x)))
None

In [9]: s.is_zero
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-fe411108cb68> in <module>
----> 1 s.is_zero

AttributeError: 'FiniteSet' object has no attribute 'is_zero'
```

asmeurer commented:
So the idea here is to leave the old assumptions machinery around on Basic in case we ever want to use it for non-Basic predicates?

I feel like it would be better to just remove the metaclass from Basic entirely. For one thing, not having the metaclass could improve performance for non-Expr classes. If we want to add old assumptions on matrices in the future (e.g.) we can just add the ManagedProperties metaclass there.

asmeurer commented:
Actually I guess we would still need to keep BasicMeta as the metaclass (it also needs to go, but that's a lot more work).

oscarbenjamin committed:
remove unwanted merge leftover

oscarbenjamin committed:
make AssumptionsWrapper subclass Expr rather than Basic

oscarbenjamin committed:
core: introduce ExprManagedProperties metaclass for Expr

oscarbenjamin committed:
add type hint for _FACTKB

oscarbenjamin committed:
Use ExprManagedProperties in quantum

oscarbenjamin commented on sympy/core/function.py:
```diff
@@ -279,6 +280,10 @@ def __repr__(cls):
 
 
+class ExprFunctionClass(FunctionClass, ExprManagedProperties):
+    pass
```
There is `BooleanFunction` which is defined as
```python
class BooleanFunction(Application, Boolean):
    pass
```
This means that `Application` cannot be an `Expr` subclass and so its metaclass `FunctionClass` must subclass `ManagedProperties`. Then `Function` is defined as
```python
class Function(Application, Expr, metaclass=ExprFunctionClass):
    pass
```
where we need this separate metaclass to distinguish from that of `BooleanFunction`.

oscarbenjamin commented:
I'm not convinced about finishing this any more. I can't really see a strong benefit in doing this any more and I'm fairly sure it would lead to minor incompatibilities with downstream code. Getting rid of the metaclasses altogether would be good but I don't see any way to do that without breaking compatibility for any downstream `Expr` subclasses that actually use the assumptions system.
